thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/HoltWintersDetector.java

```diff
@@ -0,0 +1,538 @@
+          values[i + kernelOffset] = AlgorithmUtils.robustMean(dfInput.getDoubles(COL_VALUE)
+              .slice(i, i + kernelSize), kernelSize).getDouble(kernelSize - 1);
+        }
+        dfInput.addSeries(COL_VALUE, values);
+      }
+    }
+
+    DataFrame dfCurr = new DataFrame(dfInput).renameSeries(COL_VALUE, COL_CURR);
+    DataFrame dfBase = computePredictionInterval(dfInput, window.getStartMillis()).renameSeries(COL_VALUE, COL_BASE);
+    DataFrame df = new DataFrame(dfCurr).addSeries(dfBase);
+    df.addSeries(COL_DIFF, df.getDoubles(COL_CURR).subtract(df.get(COL_BASE)));
+    df.addSeries(COL_ANOMALY, BooleanSeries.fillValues(df.size(), false));
+
+    // consistent with pattern
+    if (pattern.equals(Pattern.UP_OR_DOWN) ) {
+      df.addSeries(COL_PATTERN, BooleanSeries.fillValues(df.size(), true));
+    } else {
+      df.addSeries(COL_PATTERN, pattern.equals(Pattern.UP) ? df.getDoubles(COL_DIFF).gt(0) :
+          df.getDoubles(COL_DIFF).lt(0));
+    }
+    df.addSeries(COL_DIFF_VIOLATION, df.getDoubles(COL_DIFF).abs().gte(df.getDoubles(COL_ERROR)));
+    df.mapInPlace(BooleanSeries.ALL_TRUE, COL_ANOMALY, COL_PATTERN, COL_DIFF_VIOLATION);
+
+    // anomalies
+    List<MergedAnomalyResultDTO> result = DetectionUtils.makeAnomalies(sliceData, df, COL_ANOMALY,
+        window.getEndMillis(),
+        DetectionUtils.getMonitoringGranularityPeriod(monitoringGranularity, datasetConfig), datasetConfig);
+
+    return result;
+  }
+
+  /**
+   * Break a time period into parts and fetch it one by one, and put it together into a DataFrame
+   *
+   * @param metricEntity metric entity
+   * @param start start timestamp
+   * @param end end timestamp
+   * @return Data Frame that has data from start to end
+   */
+  private DataFrame fetchDataByPart(MetricEntity metricEntity, long start, long end) {
+    long duration = end - start;
+    List<MetricSlice> slices = new ArrayList<>();
+    long temp = start;
+    int parts = (int) (duration / PARTITION_PERIOD);
+    for (int i = 0; i < parts; i++) {
+      MetricSlice sliceData = MetricSlice.from(metricEntity.getId(), temp, temp + PARTITION_PERIOD,
+          metricEntity.getFilters(), timeGranularity);
+      temp += PARTITION_PERIOD;
+      slices.add(sliceData);
+    }
+    MetricSlice sliceData = MetricSlice.from(metricEntity.getId(), temp, end,
+        metricEntity.getFilters(), timeGranularity);
+    slices.add(sliceData);
+
+    InputData data = this.dataFetcher.fetchData(new InputDataSpec().withTimeseriesSlices(slices)
+        .withMetricIdsForDataset(Collections.singletonList(metricEntity.getId())));
+
+    LongSeries longSeries = LongSeries.buildFrom();
+    DoubleSeries doubleSeries = DoubleSeries.buildFrom();
+
+    DataFrame df = new DataFrame();
+    df.addSeries(COL_TIME, longSeries);
+    df.addSeries(COL_VALUE, doubleSeries);
+    df.setIndex(COL_TIME);
+
+    Set<Map.Entry<MetricSlice, DataFrame>> set = data.getTimeseries().entrySet();
```

Comment:
df.append() can take List<DataFrame> as param. You don't need to have iteration here.

Reply:
Not splitting request here anymore.